Performance Issues

Smartappli opened the pull request:
## Description
Please include a summary of the change, the motivation, and any additional context that will help others understand your PR. If it closes one or more open issues, [please tag them as described here](https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword).

- [x] packages/syft/src/syft/client/api.py:743:25: PERF102 When using only the values of a dict use the `values()` method
- [x] packages/syft/src/syft/client/api.py:1104:25: PERF102 When using only the values of a dict use the `values()` method
- [x] packages/syft/src/syft/client/client.py:592:44: PERF102 When using only the values of a dict use the `values()` method
- [x] packages/syft/src/syft/client/registry.py:411:17: PERF401 Use a list comprehension to create a transformed list-
- [x] packages/syft/src/syft/custom_worker/k8s.py:128:17: PERF402 Use `list` or `list.copy` to create a copy of a list
- [x] packages/syft/src/syft/custom_worker/k8s.py:156:17: PERF401 Use a list comprehension to create a transformed list
- [x] packages/syft/src/syft/node/node.py:622:29: PERF102 When using only the values of a dict use the `values()` method
- [x] packages/syft/src/syft/node/node.py:838:17: PERF401 Use a list comprehension to create a transformed list
- [x] packages/syft/src/syft/node/server.py:309:9: PERF203 `try`-`except` within a loop incurs performance overhead
- [x] packages/syft/src/syft/protocol/data_protocol.py:510:38: PERF102 When using only the keys of a dict use the `keys()` method
- [x] packages/syft/src/syft/serde/recursive.py:314:46: PERF102 When using only the values of a dict use the `values()` method
- [x] packages/syft/src/syft/serde/recursive_primitives.py:77:13: PERF401 Use a list comprehension to create a transformed list
- [x] packages/syft/src/syft/service/action/action_object.py:1097:13: PERF401 Use a list comprehension to create a transformed list
- [x] packages/syft/src/syft/service/action/action_object.py:2175:17: PERF102 When using only the values of a dict use the `values()` method
- [x] packages/syft/src/syft/service/api/api_service.py:243:13: PERF401 Use a list comprehension to create a transformed list
- [x] packages/syft/src/syft/service/code/user_code.py:445:29: PERF102 When using only the keys of a dict use the `keys()` method
- [x] packages/syft/src/syft/service/code/user_code.py:730:32: PERF102 When using only the values of a dict use the `values()` method
- [x] packages/syft/src/syft/service/code/user_code.py:879:26: PERF102 When using only the values of a dict use the `values()` method
- [x] packages/syft/src/syft/service/code_history/code_history_service.py:139:21: PERF401 Use a list comprehension to create a transformed list
- [x] packages/syft/src/syft/service/dataset/dataset.py:490:17: PERF401 Use a list comprehension to create a transformed list
- [x] packages/syft/src/syft/service/dataset/dataset_service.py:195:25: PERF401 Use a list comprehension to create a transformed list
- [x] packages/syft/src/syft/service/network/network_service.py:910:21: PERF401 Use a list comprehension to create a transformed list
- [x] packages/syft/src/syft/service/policy/policy.py:77:13: PERF401 Use a list comprehension to create a transformed list
- [x] packages/syft/src/syft/service/policy/policy.py:716:9: PERF401 Use a list comprehension to create a transformed list
- [x] packages/syft/src/syft/service/project/project.py:946:17: PERF401 Use a list comprehension to create a transformed list
- [x] packages/syft/src/syft/service/queue/zmq_queue.py:303:39: PERF102 When using only the values of a dict use the `values()` method
- [x] packages/syft/src/syft/service/queue/zmq_queue.py:479:31: PERF102 When using only the values of a dict use the `values()` method
- [x] packages/syft/src/syft/service/queue/zmq_queue.py:931:33: PERF102 When using only the values of a dict use the `values()` method
- [x] packages/syft/src/syft/service/queue/zmq_queue.py:936:32: PERF102 When using only the values of a dict use the `values()` method
- [x] packages/syft/src/syft/service/request/request.py:[125](https://github.com/OpenMined/PySyft/actions/runs/9409685607/job/25920078029?pr=8886#step:8:130)3:39: PERF102 When using only the values of a dict use the `values()` method
- [x] packages/syft/src/syft/service/user/user_roles.py:39:13: PERF401 Use a list comprehension to create a transformed list
- [x] packages/syft/src/syft/service/worker/worker_pool.py:186:17: PERF401 Use a list comprehension to create a transformed list
- [x] packages/syft/src/syft/service/worker/worker_pool.py:199:17: PERF401 Use a list comprehension to create a transformed list
- [x] packages/syft/src/syft/service/worker/worker_pool_service.py:367:13: PERF401 Use a list comprehension to create a transformed list
- [x] packages/syft/src/syft/store/blob_storage/__init__.py:140:9: PERF203 `try`-`except` within a loop incurs performance overhead
- [x] packages/syft/src/syft/store/blob_storage/seaweedfs.py:[130](https://github.com/OpenMined/PySyft/actions/runs/9409685607/job/25920078029?pr=8886#step:8:135):33: PERF203 `try`-`except` within a loop incurs performance overhead
- [x] packages/syft/src/syft/store/dict_document_store.py:83:29: PERF102 When using only the values of a dict use the `values()` method
- [x] packages/syft/src/syft/store/mongo_document_store.py:430:17: PERF401 Use a list comprehension to create a transformed list
- [x] packages/syft/src/syft/store/sqlite_document_store.py:281:13: PERF401 Use a list comprehension to create a transformed list
- [x] packages/syft/src/syft/util/schema.py:182:27: PERF102 When using only the values of a dict use the `values()` method
- [x] packages/syft/tests/syft/types/dicttuple_test.py:44:9: PERF402 Use `list` or `list.copy` to create a copy of a list
- [x] packages/syft/tests/syft/users/user_service_test.py:222:23: PERF102 When using only the keys of a dict use the `keys()` method

## Affected Dependencies
List any dependencies that are required for this change.

## How has this been tested?
- Describe the tests that you ran to verify your changes.
- Provide instructions so we can reproduce.
- List any relevant details for your test configuration.

## Checklist
- [ ] I have followed the [Contribution Guidelines](https://github.com/OpenMined/.github/blob/master/CONTRIBUTING.md) and [Code of Conduct](https://github.com/OpenMined/.github/blob/master/CODE_OF_CONDUCT.md)
- [ ] I have commented my code following the [OpenMined Styleguide](https://github.com/OpenMined/.github/blob/master/STYLEGUIDE.md)
- [ ] I have labeled this PR with the relevant [Type labels](https://github.com/OpenMined/.github/labels?q=Type%3A)
- [ ] My changes are covered by tests

Smartappli committed:
Update api.py

Smartappli committed:
Update api.py

Smartappli committed:
Update registry.py

Smartappli committed:
Update node.py

Smartappli committed:
Update k8s.py

Smartappli committed:
Update k8s.py

Smartappli committed:
Update client.py

Smartappli committed:
Update server.py

Smartappli committed:
Update server.py

Smartappli committed:
Update data_protocol.py

Smartappli committed:
Update recursive.py

Smartappli committed:
Update recursive_primitives.py

Smartappli committed:
Update recursive_primitives.py

Smartappli committed:
Update action_object.py

Smartappli committed:
Update action_object.py

Smartappli committed:
Update api_service.py

Smartappli committed:
Update api_service.py

Smartappli committed:
Update api_service.py

Smartappli committed:
Update user_code.py

Smartappli committed:
Update code_history.py

Smartappli committed:
Update dataset.py

Smartappli committed:
Update dataset_service.py

Smartappli committed:
Update dataset_service.py

Smartappli committed:
Update code_history.py

Smartappli committed:
Update dataset_service.py

Smartappli committed:
Update network_service.py

Smartappli committed:
Update network_service.py

Smartappli committed:
Update policy.py

Smartappli committed:
Update project.py

Smartappli committed:
Update project.py

Smartappli committed:
Update project.py

Smartappli committed:
Update project.py

Smartappli committed:
Update zmq_queue.py

Smartappli committed:
Update zmq_queue.py

Smartappli committed:
Update request.py

Smartappli committed:
Update user_roles.py

Smartappli committed:
Update worker_pool.py

Smartappli committed:
Update request.py

Smartappli committed:
Update worker_pool.py

Smartappli committed:
Update project.py

Smartappli committed:
Update zmq_queue.py

Smartappli committed:
Update zmq_queue.py

Smartappli committed:
Update worker_pool_service.py

Smartappli committed:
Update seaweedfs.py

Smartappli committed:
Update __init__.py

Smartappli committed:
Update dict_document_store.py

Smartappli committed:
Update mongo_document_store.py

Smartappli committed:
Update sqlite_document_store.py

Smartappli committed:
Update schema.py

Smartappli committed:
Update dicttuple_test.py

Smartappli committed:
Update user_service_test.py

Smartappli committed:
Update __init__.py

Smartappli committed:
Update seaweedfs.py

Smartappli committed:
Update mongo_document_store.py

Smartappli committed:
Update __init__.py

Smartappli committed:
Update zmq_queue.py

Smartappli committed:
Update __init__.py

Smartappli committed:
Update seaweedfs.py

Smartappli committed:
Update __init__.py

Smartappli committed:
Update zmq_queue.py

Smartappli committed:
Update dicttuple_test.py

Smartappli committed:
Update seaweedfs.py

Smartappli committed:
Update ruff.toml

Smartappli committed:
Update ruff.toml

Smartappli committed:
Update code_history_service.py

Smartappli committed:
Update code_history_service.py

Smartappli committed:
Update __init__.py

Smartappli committed:
Update __init__.py

Smartappli committed:
Update __init__.py

Smartappli committed:
Update __init__.py

Smartappli committed:
Update recursive_primitives.py

Smartappli commented:
@kiendang Can you review plz?

Smartappli committed:
Merge branch 'dev' into performance

kiendang commented on packages/syft/src/syft/node/node.py:
```diff
@@ -834,8 +834,7 @@ def __repr__(self) -> str:
         service_string = ""
         if not self.is_subprocess:
             services = []
```
This
```suggestion
```
is no longer needed here.

Smartappli replied:
done

kiendang commented on packages/syft/src/syft/service/action/action_object.py:
```diff
@@ -1090,11 +1090,9 @@ def syft_make_action(
 
-        for obj in args:
-            arg_ids.append(self._syft_prepare_obj_uid(obj))
+        kwarg_ids = {}
 
         for k, obj in kwargs.items():
             kwarg_ids[k] = self._syft_prepare_obj_uid(obj)
```
```suggestion
        kwarg_ids = {k: self._syft_prepare_obj_uid(obj) for k, obj in kwargs.items()}
```

Smartappli replied:
done. thx

kiendang commented on packages/syft/tests/syft/types/dicttuple_test.py:
```diff
@@ -39,9 +39,7 @@ def test_dict_tuple_not_subclassing_mapping():
-    values = []
-    for v in dict_tuple:
-        values.append(v)
+    values = list(dict_tuple)
```
I think here we're testing this specific syntax.

Smartappli replied:
thx

kiendang replied:
Looks like this is still not reverted. Is it auto-formatted by a linter rule? If so you can use a `# noqa: XXX` with `XXX` being the specific linter rule to prevent it from auto-formatting.

Smartappli committed:
Update node.py

Smartappli committed:
Update action_object.py

Smartappli reviewed (commented):
thx for your review

Smartappli committed:
Merge branch 'dev' into performance

Smartappli committed:
Merge branch 'dev' into performance

eelcovdw commented on packages/syft/src/syft/client/client.py:
```diff
@@ -589,7 +589,7 @@ def get_nested_codes(code: UserCode) -> list[UserCode]:
             if code.nested_codes is None:
                 return result
 
-            for _, (linked_code_obj, _) in code.nested_codes.items():
```
please check this, the changed code does not do the same thing, should be value[0] instead of key?

Smartappli replied:
thx

Smartappli committed:
Update client.py

Smartappli committed:
Merge branch 'dev' into performance

kiendang commented on packages/syft/src/syft/client/client.py:
```diff
@@ -589,7 +589,7 @@ def get_nested_codes(code: UserCode) -> list[UserCode]:
                 return result
 
-            for _, (linked_code_obj, _) in code.nested_codes.items():
+            for linked_code_obj in code.nested_codes.values():
```
```suggestion
            for (linked_code_obj, _) in code.nested_codes.values():
```

kiendang commented on packages/syft/src/syft/service/queue/zmq_queue.py:
```diff
@@ -476,8 +476,9 @@ def _run(self) -> None:
 
-            for _, service in self.services.items():
-                self.dispatch(service, None)
+            for service in self.services.keys():
```
Should this be

```suggestion
            for service in self.services.values():
```
?

Also let's keep `self.dispatch(service, None)` instead of running it in a thread here. This is a non-trivial change and should be its own PR.

Smartappli replied:
https://github.com/OpenMined/PySyft/pull/8901

kiendang commented on packages/syft/src/syft/service/worker/worker_pool.py:
```diff
@@ -194,9 +193,11 @@ def healthy_workers(self) -> list[SyftWorker] | SyftError:
         """
         _healthy_workers = []
```
This

```suggestion
```

is no longer needed

kiendang commented on packages/syft/src/syft/service/worker/worker_pool_service.py:
```diff
@@ -362,9 +362,7 @@ def get_all(
-        res: list[tuple] = []
-        for pool in worker_pools:
-            res.append((pool.name, pool))
+        res = [(pool.name, pool) for pool in worker_pools]
```
We're passing it to `DictTuple` so using a generator instead of list is a tiny bit more efficient.
```suggestion
        res = ((pool.name, pool) for pool in worker_pools)
```

kiendang commented on packages/syft/src/syft/store/blob_storage/__init__.py:
```diff
@@ -135,13 +136,13 @@ def syft_iter_content(
-                    f"Attempt {attempt}/{max_retries} failed: {e} at byte {current_byte}. Retrying..."
+                    f"Attempt {attempt + 1}/{max_retries} failed: {e} at byte {current_byte}. Retrying..."
                 )
+                time.sleep(2**attempt)  # exponential backoff
```
I think this is a potential change but let's make this exponential backoff into a separate PR.

Smartappli replied:
https://github.com/OpenMined/PySyft/pull/8900

Smartappli committed:
Update client.py

Smartappli committed:
Update zmq_queue.py

Smartappli committed:
Update worker_pool_service.py

Smartappli committed:
Update worker_pool.py

Smartappli committed:
Update __init__.py

Smartappli committed:
Update zmq_queue.py

Smartappli committed:
Update client.py

kiendang committed:
Disable linter rule for the specific syntax test

kiendang committed:
Merge branch 'dev' into performance

kiendang committed:
Disable ruff PERF rules for now

Since most PERF rules' autofixes are unsafe, adding them to run
automatically as part of pre-commit would be a bit disruptive.
Would be nice to run this from time to time, or enable them once the
autofixes are stable.

kiendang reviewed (approved):
Thanks @Smartappli. I commented out ruff PERF rules for now. Since most PERF rules' autofixes are unstable, adding them to run automatically as part of pre-commit would be a bit disruptive. Would be nice to run this from time to time, or enable them once the autofixes are stable.
